### DISCUSSION TITLE
Quickly switch between groups of settings?

### DISCUSSION THREAD

**Discussion by detly** [wezterm/wezterm#4698] [Q&A]
### What Operating System(s) are you running on?

Linux Wayland

### Which Wayland compositor or X11 Window manager(s) are you using?

Mutter

### WezTerm version

20230712-072601-f4abf8fd (flatpak)

### Ask your question!

Is there a way to quickly switch between groups of settings, ideally named in a meaningful way? Eg. I have my default settings, I have "huge" settings (for running on laptop screen/distant viewing), and I have "tiny" settings (for watching logs on portrait monitors).

In Gnome Terminal I can assign these to profiles and switch between them. It's clear that with the Lua-based config, I can have eg. a `~/.config/wezterm/wezterm.lua` (for the defaults), `tiny.lua`, etc. But switching between them is... not easy. I can do it at launch, if I want to type out the full path to the config file in my DE's keyboard shortcut, or at an existing terminal. Is there a simple way to do this from within a terminal, _for_ that terminal?

**Comment by wez** [comment]
You can use https://wezfurlong.org/wezterm/config/lua/window/set_config_overrides.html in lua to override the base config

**Reply by detly** [reply]
Sorry for the delay, I was on leave (from computers). I gave this a go, but had a couple of issues with it. Mainly: there's no separation for overrides, and there's no way to check what's been applied "as a group", if that makes sense.

That is - I have overrides for toggling the border and opacity when switching between fullscreen/windowed, very much like [the `window-resized` example](https://wezfurlong.org/wezterm/config/lua/window-events/window-resized.html?h=fullscreen#window-resized). Note that the only way to check if the settings have been applied is to pick one of the settings, inspect its value in the single config overrides table, and use that as a kind of proxy for whether the whole "fullscreen/not fullscreen" settings group has been applied.

Now if you _also_ have settings groups that apply per manual change eg. "small mode", "presentation mode", "serial" etc. and one of those groups happens to include eg. the border or opacity, they're going to clobber each other or interact poorly with the "can we stop recursing now" checks.

I mean, it's all Lua code right, a Turing-complete language, so _technically_ there's going to be a way to figure it out. But it's about how complex or prone to error that is, and the returns on effort. I wish I had time to actually make a suggestion in the form of a PR, but that's unlikely. Still, I can tell you some conspicuous things I looked for and didn't find in an effort to solve this, in case it helps:

- A lot of it boils down to just being able to use a variable in the context of what the style applies to ie. a window. This would mean that if I get the "serial mode activated" event, I can quickly tell if I'm already in fullscreen mode or not. Or even just, if I'm recursing through handling `window-resized` and `window-config-reloaded` events, knowing whether to stop or not.
 
  I notice there's a [`wezterm.GLOBAL`] store, but no analog for a window. It might be possible to use a unique identifier for a window ([`window_id()`](https://wezfurlong.org/wezterm/config/lua/window/window_id.html) maybe?) as a key for a nested table... but, complexity.

- A more specific solution might be: have a way to tell whether one or more of the event handlers is happening "recursively" or not. The word "recursively" is not very well defined there, but some of the difficulty is caused by `set_config_overrides()` emitting the `window-config-reloaded` event, so maybe having a baked-in way to detect that would help eg. an extra boolean passed to the handler. Or maybe separating the events emitted by internal vs. external config-changing events.

- Have a CSS-selector-like approach, where eg. the state change event handlers can apply selectors/tags/whatever, and something else computes settings for the current set of tags eg. `[fullscreen, serial]` to `[windowed, serial]`.

- Multiple, nameable config overrides eg. `set_config_overrides('fullscreen')`, and later `if not get_config_overrides('fullscreen').

I have not thought these through very much, they're just things I looked/wished for.
